sw/device/tests/crypto/ecdh_p256_sideload_functest.c

```diff
@@ -0,0 +1,213 @@
+  ecc_public_key_t public_keyB = {
+      .x =
+          {
+              .key_mode = kKeyModeEcdh,
+              .key_length = sizeof(xB),
+              .key = xB,
+          },
+      .y =
+          {
+              .key_mode = kKeyModeEcdh,
+              .key_length = sizeof(yB),
+              .key = yB,
+          },
+  };
+
+  // Generate a keypair.
+  LOG_INFO("Generating keypair A...");
+  TRY(otcrypto_ecdh_keygen(&kCurveP256, &private_keyA, &public_keyA));
+
+  // Generate a second keypair.
+  LOG_INFO("Generating keypair B...");
+  TRY(otcrypto_ecdh_keygen(&kCurveP256, &private_keyB, &public_keyB));
+
+  // Sanity check; public keys should be different from each other.
+  CHECK_ARRAYS_NE(xA, xB, ARRAYSIZE(xA));
+  CHECK_ARRAYS_NE(yA, yB, ARRAYSIZE(yA));
+
+  // Allocate space for two shared keys.
+  uint32_t shared_keyblobA[keyblob_num_words(kEcdhSharedKeyConfig)];
+  crypto_blinded_key_t shared_keyA = {
+      .config = kEcdhSharedKeyConfig,
+      .keyblob_length = sizeof(shared_keyblobA),
+      .keyblob = shared_keyblobA,
+      .checksum = 0,
+  };
+  uint32_t shared_keyblobB[keyblob_num_words(kEcdhSharedKeyConfig)];
+  crypto_blinded_key_t shared_keyB = {
+      .config = kEcdhSharedKeyConfig,
+      .keyblob_length = sizeof(shared_keyblobB),
+      .keyblob = shared_keyblobB,
+      .checksum = 0,
+  };
+
+  // Compute the shared secret from A's side of the computation (using A's
+  // private key and B's public key).
+  LOG_INFO("Generating shared secret (A)...");
+  TRY(otcrypto_ecdh(&private_keyA, &public_keyB, &kCurveP256, &shared_keyA));
+
+  // Compute the shared secret from B's side of the computation (using B's
+  // private key and A's public key).
+  LOG_INFO("Generating shared secret (B)...");
+  TRY(otcrypto_ecdh(&private_keyB, &public_keyA, &kCurveP256, &shared_keyB));
+
+  // Get pointers to individual shares of both shared keys.
+  uint32_t *keyA0;
+  uint32_t *keyA1;
+  TRY(keyblob_to_shares(&shared_keyA, &keyA0, &keyA1));
+  uint32_t *keyB0;
+  uint32_t *keyB1;
+  TRY(keyblob_to_shares(&shared_keyB, &keyB0, &keyB1));
+
+  // Unmask the keys and check that they match.
+  uint32_t keyA[kP256CoordWords];
+  uint32_t keyB[kP256CoordWords];
+  for (size_t i = 0; i < ARRAYSIZE(keyA); i++) {
+    keyA[i] = keyA0[i] ^ keyA1[i];
+    keyB[i] = keyB0[i] ^ keyB1[i];
+  }
+  CHECK_ARRAYS_EQ(keyA, keyB, ARRAYSIZE(keyA));
+
+  return OTCRYPTO_OK;
+}
+
+static status_t test_setup(void) {
+  // Initialize the key manager and advance to OwnerRootKey state.
+  dif_keymgr_t keymgr;
+  dif_kmac_t kmac;
+  TRY(keymgr_testutils_startup(&keymgr, &kmac));
+  TRY(keymgr_testutils_advance_state(&keymgr, &kOwnerIntParams));
+  TRY(keymgr_testutils_advance_state(&keymgr, &kOwnerRootKeyParams));
+  TRY(keymgr_testutils_check_state(&keymgr, kDifKeymgrStateOwnerRootKey));
+
+  // Initialize entropy complex, which the key manager uses to clear sideloaded
+  // keys. The `keymgr_testutils_startup` function restarts the device, so this
```

Comment:
nit: I believe the `keymgr` also uses the entropy complex to mix in additional entropy in the key derivation scheme, so in the `ROM_EXT` the expected use case is that the entropy complex will be initialized before advancing the state of the `keymgr`.

Reply:
Right now `keymgr_testutils` configures KMAC to use software entropy in order to avoid depending on the entropy complex; does that apply to keymgr as well or is it a different setting?

Reply:
That is part of it. There may be additional entropy requirements inside the `keymgr` that we'll have to check at some point. 

Given that there is limited boot entropy available, in production we'll most likely have to switch the entropy complex to continuous mode before advancing the `keymgr`. I wanted to bring this up given that the comment can be misinterpreted by a reader not familiar with the entropy complex quirks.

Reply:
I've updated the comment (here and in the other 2 sideloaded-key tests where it appears) to explain that the keymgr is initialized here using software entropy by the testutils, but that this isn't the expected case in production, and to describe the call to `entropy_complex_init()` as "initializing the entropy complex for cryptolib".